test/dns/system_resolver.cpp

```diff
@@ -5,159 +5,119 @@
 #define CATCH_CONFIG_MAIN
 #include "../src/libmeasurement_kit/ext/catch.hpp"
 
-#include "../src/libmeasurement_kit/dns/system_resolver_impl.hpp"
+#include "../src/libmeasurement_kit/dns/system_resolver.hpp"
 
 using namespace mk;
 using namespace mk::dns;
 
-static const char *null_inet_ntop(int, const void *, char *, socklen_t) {
+static const char *fail_inet_ntop(int, const void *, char *, socklen_t) {
     return nullptr;
 }
 
-static int null_getaddrinfo(const char *, const char *, const struct addrinfo *,
+static int fail_getaddrinfo(const char *, const char *, const struct addrinfo *,
                             struct addrinfo **) {
     return EAI_FAIL;
 }
 
-TEST_CASE("the system resolver can handle a getaddrinfo error") {
-    loop_with_initial_event([]() {
-        system_resolver_impl<null_getaddrinfo>(
-            "IN", "A", "www.neubot.org",
-            [](Error e, Var<Message>) {
-                REQUIRE(e == NonRecoverableFailureError());
-                break_loop();
-            },
-            {}, Reactor::global(), Logger::global());
+template <MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+void run_system_resolver(std::string dns_class, std::string dns_type,
+                         std::string name, Callback<Error, Var<Message>> cb) {
+    Var<Reactor> reactor = Reactor::make();
+    Var<Logger> logger = Logger::make();
+    reactor->run_with_initial_event([&]() {
+        system_resolver<getaddrinfo, inet_ntop>(dns_class, dns_type, name,
+                                                reactor, logger,
+                                                [&](Error e, Var<Message> m) {
+                                                    reactor->stop();
+                                                    cb(e, m);
+                                                });
     });
 }
 
+TEST_CASE("the system resolver can handle a getaddrinfo error") {
+    run_system_resolver<fail_getaddrinfo>(
+        "IN", "A", "www.neubot.org", [](Error e, Var<Message>) {
+            REQUIRE(e == NonRecoverableFailureError());
+        });
+}
+
 TEST_CASE("the system resolver can handle a inet_ntop error") {
-    loop_with_initial_event([]() {
-        system_resolver_impl<getaddrinfo, null_inet_ntop>(
-            "IN", "A", "neubot.org",
-            [](Error e, Var<Message>) {
-                REQUIRE(e == InetNtopFailureError());
-                break_loop();
-            },
-            {}, Reactor::global(), Logger::global());
-    });
+    run_system_resolver<getaddrinfo, fail_inet_ntop>(
+        "IN", "A", "neubot.org",
```

Comment:
Since we don't own this domain maybe we oughta use something else here. Would it be useful to setup some A records on `ooni.io` just for doing CI?